### ISSUE TITLE
Database corruption renders extension unusable with no obvious way to fix it

### ISSUE THREAD

**Issue by piousdeer** [SUPERCILEX/gnome-clipboard-history#96]
### Describe the bug

My clipboard history has suddenly stopped working, always displaying a blank list. After many days of thinking my distro was to blame, I was finally motivated enough to debug it. My database.log was somehow corrupted: at some point in time, `OP_TYPE_SAVE_TEXT <new entry text>, OP_TYPE_DELETE_TEXT <oldest entry id>` started to get written to the file not once, but twice every time I copied something. Since then, whenever the extension was enabled, it crashed [on this line](https://github.com/SUPERCILEX/gnome-clipboard-history/blob/f12fd6725ade9bf3f737b81f65a9ebd00af1b561/store.js#L144) after trying to delete an entry which was already deleted.

I was unable to reproduce the bug and I don't know how to prevent it, but I suggest the following changes:
- Make the extension more resilient to this kind of data corruption (log warnings instead of crashing)
- In case of a critical error, notify the user and offer them to report a bug and reset the DB

#### Versions

```
GNOME Shell 42.4
GCH   Version: 15
org.gnome.shell.extensions.clipboard-history cache-only-favorites false
org.gnome.shell.extensions.clipboard-history cache-size 100
org.gnome.shell.extensions.clipboard-history clear-history @as []
org.gnome.shell.extensions.clipboard-history confirm-clear true
org.gnome.shell.extensions.clipboard-history disable-down-arrow false
org.gnome.shell.extensions.clipboard-history display-mode 2
org.gnome.shell.extensions.clipboard-history enable-keybindings true
org.gnome.shell.extensions.clipboard-history history-size 10
org.gnome.shell.extensions.clipboard-history move-item-first true
org.gnome.shell.extensions.clipboard-history next-entry @as []
org.gnome.shell.extensions.clipboard-history notify-on-copy false
org.gnome.shell.extensions.clipboard-history paste-on-selection true
org.gnome.shell.extensions.clipboard-history prev-entry @as []
org.gnome.shell.extensions.clipboard-history private-mode false
org.gnome.shell.extensions.clipboard-history process-primary-selection false
org.gnome.shell.extensions.clipboard-history strip-text false
org.gnome.shell.extensions.clipboard-history toggle-menu ['<Super>V']
org.gnome.shell.extensions.clipboard-history toggle-private-mode ['<Super><Shift>P']
org.gnome.shell.extensions.clipboard-history topbar-preview-size 10
org.gnome.shell.extensions.clipboard-history window-width-percentage 26
```

#### Steps to repro

I couldn't find out what caused the corruption, but I could replicate it by modifying the code in a such way that `enable()` always gets called twice.

### Additional context (if a crash, provide stack trace)

```
Aug 20 18:21:24 leginix .gnome-shell-wr[2247]: JS ERROR: TypeError: undefined has no properties
                                               loop/<@/home/v/.nix-profile/share/gnome-shell/extensions/clipboard-history@alexsaveau.dev/store.js:144:56
                                               parseAvailableAware@/home/v/.nix-profile/share/gnome-shell/extensions/clipboard-history@alexsaveau.dev/store.js:189:7
                                               loop@/home/v/.nix-profile/share/gnome-shell/extensions/clipboard-history@alexsaveau.dev/store.js:142:26
                                               loop/<@/home/v/.nix-profile/share/gnome-shell/extensions/clipboard-history@alexsaveau.dev/store.js:137:11
```

**Comment by SUPERCILEX** [comment]
There have actually been other reports tracing back to the same issue: https://github.com/SUPERCILEX/gnome-clipboard-history/issues/67 and another one I'm too lazy to look for.

> Make the extension more resilient to this kind of data corruption (log warnings instead of crashing)

Definitely not. The whole point is that if things are corrupted I have no idea why and can't fix it, otherwise I wouldn't crash. A very interesting idea you suggested though is to do something like move the old database file to a backup and start with a fresh database that has a single item telling the user to report a bug and what happened. That said, this is the last corruption bug I've known about, so probs not worth investing in that infrastructure.

Anyway, your point about enable being called twice actually makes total sense. You would end up with two copies of the extension running at the same time and bad things would happen. I think this bug can be fixed very easily by just checking if we've already initialized.

---

TL;DR: This is great! And exactly why crashing is good: I pissed you off enough that you wasted your time investigating this thing and now it should be easy to fix. :)

**Comment by piousdeer** [comment]
> The whole point is that if things are corrupted I have no idea why and can't fix it, otherwise I wouldn't crash

You're right, that wasn't a great idea.

> this is the last corruption bug I've known about, so probs not worth investing in that infrastructure

I disagree: I think it'd be worth it even if there were no bugs that could cause a corruption (and they wouldn't appear in the future), because software shouldn't fail silently, especially due to a malformed file in ~/.cache. Plus that might have caused and may cause less determined users to give up before they fix it and/or file an issue.

> You would end up with two copies of the extension running at the same time and bad things would happen. I think this bug can be fixed very easily by just checking if we've already initialized.

I'm not sure if that's what happened though, and if it is, it could instead be something for GNOME to address. I'll leave it for you to decide if that should be added 🙂

**Comment by SUPERCILEX** [comment]
> I disagree: I think it'd be worth it even if there were no bugs that could cause a corruption (and they wouldn't appear in the future), because software shouldn't fail silently, especially due to a malformed file in ~/.cache. Plus that might have caused and may cause less determined users to give up before they fix it and/or file an issue.

That's a great point. I won't have time to work on this soon, but I created #97 to track this.

> I'm not sure if that's what happened though, and if it is, it could instead be something for GNOME to address.

Yeah, I'm adding an initialization check that I hope will help, but it's perfectly possible that there are multiple gnome shell instances running. That one would be much more difficult to work around, though should be possible with a lock file.

**Comment by darkdragon-001** [comment]
I had a similar issue. I copied a lot of source code and console logs and probably some entry couldn't be parsed correctly. This made this extension unusable without any hint what could be the problem.

I saved a backup of the problematic `backup.log` but can't share it for privacy reasons. If you tell me how I could help analyzing this bug, I would be happy to assist.

**Comment by SUPERCILEX** [comment]
What version of the extension are you using? And you can dump the db metadata with this: https://github.com/SUPERCILEX/gnome-clipboard-history/issues/67#issuecomment-1133395584

**Comment by darkdragon-001** [comment]
Damned, I finally wanted to have a look but just noticed that I had cleaned the `~/.cache` directory before...  I will look out if this happens again!
I am using Fedora 36 at the moment.with Clipboard History 21 (try to keep both up to date).

**Comment by SUPERCILEX** [comment]
Dang, well let me know if it happens again. :+1: And please just `cp ~/.cache/clipboard-history@alexsaveau.dev/database.log` somewhere if so. :) Even if you don't want to give it to me, at least we can do a bit of analysis on it.
